src/Spryker/Zed/SynchronizationBehavior/Persistence/Propel/Behavior/SynchronizationBehavior.php

```diff
@@ -1197,4 +1200,129 @@ public function getTableOrFail(): Table
 
         return $table;
     }
+
+    /**
+     * @return int
+     */
+    protected function isDirectSyncPerEntityDisabled(): int
+    {
+        return isset($this->getParameters()['direct_sync_disabled']) ? 1 : 0;
+    }
+
+    /**
+     * @return string
+     */
+    protected function addIsDirectSyncEnabledMethod(): string
+    {
+        $isDirectSynchronizationEnabled = $this->getConfig()->isDirectSynchronizationEnabled() ? 1 : 0;
+        $isDirectSyncPerEntityDisabled = $this->isDirectSyncPerEntityDisabled();
+
+        return "
+/**
+ * @return int
+ */
+protected function isDirectSyncEnabled(): int
+{
+    return $isDirectSynchronizationEnabled && !$isDirectSyncPerEntityDisabled;
+}
+        ";
+    }
+
+    /**
+     * @return string
+     */
+    protected function addSendMessageMethod(): string
+    {
+        return "
+/**
+ * @param array \$message
+ *
+ * @return void
+ */
+protected function sendMessage(array \$message): void
+{
+    if (\$this->_locator === null) {
+        \$this->_locator = \\Spryker\\Zed\\Kernel\\Locator::getInstance();
+    }
+
+    \$synchronizationFacade = \$this->_locator->synchronization()->facade();
+    if (\$this->isDirectSyncEnabled() && method_exists(\$synchronizationFacade, 'addSynchronizationMessageToBuffer')) {
+        \$this->sendToBuffer(\$message);
+
+        return;
+    }
+
+    \$this->sendToQueue(\$message);
+}
+        ";
+    }
+
+    /**
+     * @throws \Spryker\Zed\SynchronizationBehavior\Persistence\Propel\Behavior\Exception\InvalidConfigurationException
+     *
+     * @return string
+     */
+    protected function addSendToBufferMethod(): string
+    {
+        $queueName = $this->getParameter('queue_group')['value'];
+        $resource = $this->getParameter('resource')['value'];
+        $queuePoolName = $this->getQueuePoolName();
+        $hasStore = $this->hasStore();
+        $hasLocale = $this->hasLocale();
+
+        if ($hasStore && $queuePoolName) {
+            throw new InvalidConfigurationException(
+                sprintf(static::ERROR_MUTUALLY_EXCLUSIVE_PARAMETERS, $this->getTableOrFail()->getPhpName()),
+            );
+        }
+
+        $setLocale = $hasLocale ? '$synchronizationMessageTransfer->setLocale($this->locale);' : '';
+
+        $setMessageQueueRouting = '';
+        if ($hasStore) {
+            $setMessageQueueRouting = '$queueSendTransfer->setStoreName($this->store);';
+        }
+
+        if ($queuePoolName) {
+            $setMessageQueueRouting = "\$queueSendTransfer->setQueuePoolName('$queuePoolName');";
+        }
+
+        if ($queueName === null) {
+            $queueName = $resource;
+        }
```

Comment:
looks like code duplication can we move this to the separate method with the meaningful name?